### PR TITLE
Fix for non-utf-8 headers in ports

### PR DESCRIPTION
**Pull request by pmp-p**
utils.read_file can only read unicode files but eg SDL2_gfx has a non unicode header "SDL2_gfxPrimitives_font.h"

### CODE REVIEW

**Comment by sbc100** [comment]
What is the "check operation" you are referring too and what is the error you are seeing?

**Comment by pmp-p** [comment]
the check is `utils.read_file(src) == utils.read_file(dest)`
and read_file only read utf-8 files https://github.com/emscripten-core/emscripten/blob/aa250479f654fb8dffb0ff69132da1779128da99/tools/utils.py#L63

**Comment by sbc100** [comment]
> the check is `utils.read_file(src) == utils.read_file(dest)` and `utils.read_file()` only read utf-8 files
> 
> https://github.com/emscripten-core/emscripten/blob/aa250479f654fb8dffb0ff69132da1779128da99/tools/utils.py#L63
> 
> read SDL2_gfxPrimitives_font.h with it raises a decoding error, which is harmless since file is already copied in that case.
> 
> can probably be repro with `embuild --pic build sdl2;reset;embuild --pic build sdl2_gfx` ( reset to remove the clutter of sdl2 build )

That seems to work fine for me:

```
$ ./embuilder --pic build sdl2
embuilder:INFO: building sdl2
ports:INFO: retrieving port: sdl2 from https://github.com/libsdl-org/SDL/archive/release-2.24.2.zip
ports:INFO: unpacking port: sdl2
cache:INFO: generating port: sysroot/lib/wasm32-emscripten/pic/libSDL2.a... (this will be cached in "/usr/local/google/home/sbc/dev/wasm/emscripten/cache/sysroot/lib/wasm32-emscripten/pic/libSDL2.a" for subsequent builds)
system_libs:INFO: compiled 115 inputs in 3.01s
cache:INFO:  - ok
embuilder:INFO: ...success. Took (5.41s)
$ ./embuilder --pic build sdl2_gfx
embuilder:INFO: building sdl2_gfx
ports:INFO: retrieving port: sdl2_gfx from https://github.com/svn2github/sdl2_gfx/archive/2b147ffef10ec541d3eace326eafe11a54e635f8.zip
ports:INFO: unpacking port: sdl2_gfx
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/pic/libSDL2_gfx.a... (this will be cached in "/usr/local/google/home/sbc/dev/wasm/emscripten/cache/sysroot/lib/wasm32-emscripten/pic/libSDL2_gfx.a" for subsequent builds)
system_libs:INFO: compiled 4 inputs in 0.90s
cache:INFO:  - ok
embuilder:INFO: ...success. Took (2.25s)
```

Are you able to repo locally?

**Comment by sbc100** [comment]
> the check is `utils.read_file(src) == utils.read_file(dest)` and `utils.read_file()` 

Does that check actually run anywhere?  How did you run across this failure?

**Comment by pmp-p** [comment]
locally (pclinuxos python3.10 ) and github CI http://pmp-p.ddns.net/paste/screen/shot-2023-11-06_1699247352.png

**Comment by sbc100** [comment]
Do you know why it happens for you can not for me?   Any idea why I can't reproduce it?   (I'm asking mostly so I can re-create a test case that we can lang alongside the fix).

**Comment by pmp-p** [comment]
Well i know for sure that the python code logic is wrong because nothing says C header cannot be latin-1, but why it is being executed in these particular case i have no idea :D

i thought originally it was because my i18n on my dev pc, but that makes litle sense on the github CI and i always put LC_ALL=C in all my scripts anyway.

**Comment by sbc100** [comment]
> Well i know for sure that the python code logic is wrong because nothing says C header cannot be latin-1, but why it is being executed in these particular cases i have no idea :D
> 
> i thought originally it was because my i18n on my dev pc, but that makes little sense on the github CI and i always put LC_ALL=C in all my scripts anyway.

I guess it could be a python-version dependent issue?    But I'm surprised nobody else has run into it.  I guess not many folk use that particular port.

File: tools/ports/__init__.py [hunk 1]
```diff
@@ -90,8 +90,11 @@ def maybe_copy(src, dest):
   headers will be "re-installed" but we skip the actual filesystem mods
   to avoid racing with other processes that might be reading these files.
   """
-  if os.path.exists(dest) and utils.read_file(src) == utils.read_file(dest):
-    return
+  try:
+    if os.path.exists(dest) and utils.read_file(src) == utils.read_file(dest):
```
Comment (sbc100) [inline tools/ports/__init__.py:94]: How about we switch to `utils.read_binary` here instead?

Comment (pmp-p) [reply tools/ports/__init__.py]: indeed, that will save a copy operation when it happens to be binary

File: tools/ports/__init__.py [hunk 2]
```diff
@@ -89,8 +89,8 @@ def maybe_copy(src, dest):
   building/installing a new flavor of a given library.  In this case the
   headers will be "re-installed" but we skip the actual filesystem mods
   to avoid racing with other processes that might be reading these files.
-  """
-  if os.path.exists(dest) and utils.read_file(src) == utils.read_file(dest):
+  """
```
Comment (sbc100) [inline tools/ports/__init__.py:92]: Looks like some whitespace was accidentally added here maybe?

**Review by sbc100 — COMMENTED** [review]
lgtm % comment

**Review by sbc100 — APPROVED** [review]

**Pull request merged**